### PR TITLE
Improved output of formatNumberReadable

### PR DESCRIPTION
**Pull request by Ruko97**
Follow up to issue #9601 and pull request #13163.
Based on advice by @cameel in the conversation in issue.
Previously `formatNumberReadable(57896044618658097711785492504343953926634992332820282019728792003956564819968) `would equal to `0x80 * 2 ** 248`, but now it equals to `2 ** 255`.
Also, there was a bug in the test case `test_format_number_readable_signed` in `test/StringUtils.cpp`. `formatNumberReadable(-0xffffffff)` should be equal to `-2 ** 32 + 1` but previously it was tested to be equal to `-2 ** 32 - 1`. The new version of `formatNumberReadable` fixes this problem.

### CODE REVIEW

File: libsolutil/StringUtils.h [hunk 1]
```diff
@@ -101,109 +101,122 @@ std::string joinHumanReadablePrefixed
...
 {
 	static_assert(
-		std::is_same<bigint, T>::value || !std::numeric_limits<T>::is_signed,
-		"only unsigned types or bigint supported"
-	); //bigint does not carry sign bit on shift
+		std::numeric_limits<T>::is_integer,
+		"only integer numbers are supported"
+	);
+
+	bool _isNegative = false;
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]: Names starting with leading underscore are used for[ parameter variables](https://github.com/ethereum/solidity/blob/develop/CODING_STYLE.md#4-variable-prefixes). Please remove `_` from here and also `_sign` below so it is clear they are local.

File: libsolutil/StringUtils.h [hunk 2]
```diff
@@ -101,109 +101,122 @@ std::string joinHumanReadablePrefixed
...
 		return _value.str();
 
+	if (_isNegative && bigint(-1 * _value) <= 0x1000000)
+		return "-" + (bigint(-1 * _value)).str();
+
 	HexCase hexcase = HexCase::Mixed;
-	HexPrefix prefix = HexPrefix::Add;
+	HexPrefix hexprefix = HexPrefix::Add;
+
+	std::string _sign = _isNegative ? "-" : "";
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]: Same as before regarding underscore. 
Also, maybe we could declare and define this before so it can be used in the previous `return` values.

File: libsolutil/StringUtils.h [hunk 3]
```diff
@@ -101,109 +101,122 @@ std::string joinHumanReadablePrefixed
...
 	static_assert(
-		std::is_same<bigint, T>::value || !std::numeric_limits<T>::is_signed,
-		"only unsigned types or bigint supported"
-	); //bigint does not carry sign bit on shift
+		std::numeric_limits<T>::is_integer,
+		"only integer numbers are supported"
+	);
+
+	bool _isNegative = false;
+	bigint v;
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
	bigint signedValue = isNegative ? (bigint(-1) * _value) : bigint(_value);
	bigint v;
```
What do you think about using `v` as the working var used by the `for` loops? It can then be initialized with `signedValue` inside each `for` instead of being assigned the ternary operator result before the loops.
Also, `signedValue` can be used in the `return` values.

**Review by matheusaaguiar — CHANGES_REQUESTED** [review]
Looks good!
Just some code style adjustments and readability suggestions.

File: libsolutil/StringUtils.h (hunk 2 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h:141]: I defined `sign` after the previous return values because the condition for entering the return statement was to have `isNegative` be true. So there wasn't any need to use the value of sign instead of just adding `"-"` before the value

File: libsolutil/StringUtils.h (hunk 3 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: Yes, I think that would be a great idea.

File: libsolutil/StringUtils.h [hunk 4]
```diff
@@ -101,109 +101,117 @@ std::string joinHumanReadablePrefixed
...
-		std::is_same<bigint, T>::value || !std::numeric_limits<T>::is_signed,
-		"only unsigned types or bigint supported"
-	); //bigint does not carry sign bit on shift
+		std::numeric_limits<T>::is_integer,
+		"only integer numbers are supported"
+	);
+
+	bool isNegative = _value < 0;
+	bigint temp;
+	bigint absoluteValue = isNegative ? (bigint(-1) * _value) : bigint(_value);
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]: Maybe a little nitpick, but it seems to me that `absoluteValue` means how far from zero the number is and then it cannot be negative. What about `decimalValue` (or `signedValue` as suggested before)?

File: libsolutil/StringUtils.h [hunk 5]
```diff
@@ -101,109 +101,117 @@ std::string joinHumanReadablePrefixed
...
 
-	// when multiple trailing zero bytes, format as N * 2**x
+	// when multiple trailing zero bytes, format as N * 2**x (if positive)
+	// or -N * 2**x (if negative)
 	int i = 0;
-	T v = _value;
-	for (; (v & 0xff) == 0; v >>= 8)
+	temp = absoluteValue;
+
+	for (; (temp & 0xff) == 0; temp >>= 8)
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
	for (temp = absoluteValue; (temp & 0xff) == 0; temp >>= 8)
```
This is more compact and has the same readability.

File: libsolutil/StringUtils.h [hunk 6]
```diff
@@ -101,109 +101,117 @@ std::string joinHumanReadablePrefixed
...
 	}
 
-	// when multiple trailing FF bytes, format as N * 2**x - 1
+	// when multiple trailing FF bytes, format as N * 2**x - 1 (if positive)
+	// or -N * 2**x + 1 (if negative)
 	i = 0;
-	for (v = _value; (v & 0xff) == 0xff; v >>= 8)
+	temp = absoluteValue;
+
+	for (; (temp & 0xff) == 0xff; temp >>= 8)
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
	for (temp = absoluteValue; (temp & 0xff) == 0xff; temp >>= 8)
```

File: libsolutil/StringUtils.h (hunk 4 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: I think signedValue would be better

**Comment by matheusaaguiar** [comment]
I think this is in good shape. Please, rebase.

**Comment by Ruko97** [comment]
Pushed a rebased version right now. Please check it.

File: libsolutil/StringUtils.h [hunk 7]
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
 		return _separator + joinHumanReadable(_list, _separator, _lastSeparator);
 }
 
-/// Same as @ref formatNumberReadable but only for unsigned numbers
+/// Formats large numbers to be easily readable by humans.
+/// Returns decimal representation for smaller numbers; hex for large numbers.
+/// "Special" numbers, powers-of-two and powers-of-two minus 1, are returned in
+/// formulaic form like 0x01 * 2**24 - 1.
+/// @a T can be any integer variable, will typically be u160, u256 or bigint.
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
/// @a T can be any integer type, will typically be u160, u256 or bigint.
```
Ignore upper case letter `A`, it should be `a`, but github is somehow tagging a user there... 
Just to be clear, suggestion is to change `variable` to `type`

File: libsolutil/StringUtils.h [hunk 8]
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
...
+/// @param _value to be formatted
+/// @param _useTruncation if true, internal truncation is also applied,
+/// like  0x5555...{+56 more}...5555
+/// @example formatNumberReadable((u256)0x7ffffff) = "0x08 * 2**24"
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
/// @example formatNumberReadable((u256)0x7ffffff) = "2**27 - 1"
```
Shouldn't this have the same value as in the check at line 119 of `test/libsolutil/StringUtils.cpp` ?

File: libsolutil/StringUtils.h [hunk 9]
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
...
 	T const& _value,
 	bool _useTruncation = false
 )
 {
 	static_assert(
-		std::is_same<bigint, T>::value || !std::numeric_limits<T>::is_signed,
-		"only unsigned types or bigint supported"
-	); //bigint does not carry sign bit on shift
+		std::numeric_limits<T>::is_integer,
+		"only integer numbers are supported"
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
		"only integer types are supported"
```

File: libsolutil/StringUtils.h [hunk 10]
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
...
+	// or -N * 2**x + 1 (if negative)
 	i = 0;
-	for (v = _value; (v & 0xff) == 0xff; v >>= 8)
+	for (temp = signedValue; (temp & 0xff) == 0xff; temp >>= 8)
 		++i;
 	if (i > 2)
 	{
+		std::string suffix = isNegative ? " + 1" : " - 1";
+
 		// 0xFF yields 2**8 - 1 (v is 0 in that case)
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
		// 0xFF yields 2**8 - 1 (temp is 0 in that case)
```
Also, I guess we could remove `(temp is 0 in that case)`.

File: libsolutil/StringUtils.h [hunk 11]
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
...
+				" * 2**" + std::to_string(i * 8) +
+				suffix;
 	}
 
-	std::string str = toHex(toCompactBigEndian(_value), prefix, hexcase);
+	std::string str = toHex(toCompactBigEndian(signedValue), hexprefix, hexcase);
+
 	if (_useTruncation)
 	{
 		// return as interior-truncated hex.
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]: I'd say that the `if` line already makes this clear. I think the comment can be removed.

File: libsolutil/StringUtils.h [hunk 12]
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
...
-		return str.substr(0, initialChars) +
+		return sign +
+			str.substr(0, initialChars) +
 			"...{+" +
 			std::to_string(numSkipped) +
 			" more}..." +
 			str.substr(len-finalChars, len);
 	}
 
 	// otherwise, show whole value.
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]: I think this is redundant too.

File: libsolutil/StringUtils.h [hunk 13]
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
...
+	if (signedValue <= 0x1000000)
+		return sign + signedValue.str();
 
 	HexCase hexcase = HexCase::Mixed;
-	HexPrefix prefix = HexPrefix::Add;
+	HexPrefix hexprefix = HexPrefix::Add;
 
-	// when multiple trailing zero bytes, format as N * 2**x
+	// when multiple trailing zero bytes, format as N * 2**x (if positive)
+	// or -N * 2**x (if negative)
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
	// when multiple trailing zero bytes, format as N * 2**x
```
I think that distinction is handled by `sign` and only relevant at the `return`, so no need to have a multiple line comment.

**Review by matheusaaguiar — CHANGES_REQUESTED** [review]
Sorry, I had another look and found some things that were glossed over before and could be improved.

**Comment by leonardoalt** [comment]
Auto please rebase onto origin/develop to get rid of the CI errors.

File: libsolutil/StringUtils.h (hunk 10 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h:170]: It could help other programmers update the code.

File: libsolutil/StringUtils.h (hunk 13 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: I wanted to write `format as ±N * 2**x` in the comment here to indicate that the printed variable is shown with its sign, but I wasn't sure if Unicode is supported in comments.

**Comment by Ruko97** [comment]
I have rebased and updated my commit based on the suggestions here. Sorry for being late.

File: libsolutil/StringUtils.h [hunk 14]
```diff
@@ -101,109 +101,111 @@ std::string joinHumanReadablePrefixed
 		return _separator + joinHumanReadable(_list, _separator, _lastSeparator);
 }
 
-/// Same as @ref formatNumberReadable but only for unsigned numbers
+/// Formats large numbers to be easily readable by humans.
+/// Returns decimal representation for smaller numbers; hex for large numbers.
+/// "Special" numbers, powers-of-two and powers-of-two minus 1, are returned in
+/// formulaic form like 0x01 * 2**24 - 1.
+/// @a T can be any integer value, will typically be u160, u256 or bigint.
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]: I don't want to be pedantic but it looks to me that `T` refers to a type like `u160` or `bigint`. It is the function param variable   `_value` that stores the integer value of type `T`.
```suggestion
/// @a T can be any integer type, will typically be u160, u256 or bigint.
```

File: libsolutil/StringUtils.h (hunk 13 above)
Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]: Right, fair enough. I've noticed that also that you changed the templates to `N * 2**x - 1` and `-N * 2**x + 1` , so it seems worth having a comment about this.

File: libsolutil/StringUtils.h (hunk 11 above)
Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]: What about this and the next comment? I think that the code already makes it explicit, so we could remove this line.
```suggestion

```

File: libsolutil/StringUtils.h (hunk 12 above)
Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]:
```suggestion

```

File: libsolutil/StringUtils.h (hunk 14 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: Sorry, I accidentally might have forgot to add this edit to my code. My bad. It's fixed now.

Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]: No problem :)

File: libsolutil/StringUtils.h (hunk 11 above)
Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]: @Ruko97 , what do you think about this?

File: libsolutil/StringUtils.h (hunk 12 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: Fixed now

File: libsolutil/StringUtils.h (hunk 11 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: This comment existed in the prior version of the function, I just left it as it is.

**Comment by matheusaaguiar** [comment]
> I have rebased and updated my commit based on the suggestions here. Sorry for being late.

No need to apologize, you are not late at all. You are actually quite timely with your responses. Thanks for contributing to the project :)

**Review by matheusaaguiar — COMMENTED** [review]
I consider it ready. But it needs another rebase on top of `develop` since we are having a release today.

**Review by matheusaaguiar — DISMISSED** [review]
It is going to need another rebase, but I am approving so someone else from the team can merge it later.

File: libsolutil/StringUtils.h [hunk 15]
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
...
 	if (i > 2)
 	{
 		// 0x100 yields 2**8 (N is 1 and redundant)
-		if (v == 1)
-			return "2**" + std::to_string(i * 8);
-		return toHex(toCompactBigEndian(v), prefix, hexcase) +
-			" * 2**" +
-			std::to_string(i * 8);
+		if (temp == 1)
+			return sign + "2**" + std::to_string(i * 8);
```
Comment (nikola-matic) [inline libsolutil/StringUtils.h]: Hey, sorry for being a little late to the party. Can you please swap all instances of string concatenation with `fmt::format`. 
E.g., in this case, it would be something like this:
```
return fmt::format("{}2**{}", sign, std::to_string(i * 8);
```
Otherwise, looks good!

File: libsolutil/StringUtils.h [hunk 16]
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
...
+			return sign + "2**" + std::to_string(i * 8 + j);
+		}
+		else
+			return sign +
+				toHex(toCompactBigEndian(temp), hexprefix, hexcase) +
+				" * 2**" +
+				std::to_string(i * 8);
 	}
 
 	// when multiple trailing FF bytes, format as N * 2**x - 1
```
Comment (nikola-matic) [inline libsolutil/StringUtils.h]: Shouldn't this now be commented as `2**x  - 1` instead of `N * 2**x  - 1`? Or at least that `N` is optional depending on whether the number is a power of 2?

File: test/libsolutil/StringUtils.cpp [hunk 17]
```diff
@@ -190,7 +206,7 @@ BOOST_AUTO_TEST_CASE(test_format_number_readable_signed)
 
 	BOOST_CHECK_EQUAL(formatNumberReadable(b, true), "-0x5555...{+56 more}...5555");
 	BOOST_CHECK_EQUAL(formatNumberReadable(c, true), "-0x0BCD...{+56 more}...6789");
-	BOOST_CHECK_EQUAL(formatNumberReadable(d, true), "-0x5555555555555556 * 2**192 - 1");
+	BOOST_CHECK_EQUAL(formatNumberReadable(d, true), "-0x5555555555555556 * 2**192 + 1");
```
Comment (matheusaaguiar) [inline test/libsolutil/StringUtils.cpp:265]: Does this change has to do with the bug you mentioned, as in
> `formatNumberReadable(-0xffffffff)` should be equal to `-2 ** 32 + 1` but previously it was tested to be equal to `-2 ** 32 - 1`

?

Also, it is a bit difficult to track what these hex values translate to as decimals. How are you making sure that this is indeed the right value?

File: libsolutil/StringUtils.h (hunk 16 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: No, not always. `N` can be unequal to 1 if the variable is not exactly a power of 2 or one-less-than a power of 2. For example, `9 * 2**249` would be displayed as `18 * 2**248`.

File: test/libsolutil/StringUtils.cpp (hunk 17 above)
Comment (Ruko97) [reply test/libsolutil/StringUtils.cpp:265]: Yes, the change is related to the bug I mentioned
Much of the test values is carried from previous values that were being tested. I only changed some of the variables to negative. I also used `isoltest` to verify the values.

File: libsolutil/StringUtils.h (hunk 16 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: Perhaps I should add a test regarding this.

File: libsolutil/StringUtils.h (hunk 15 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: ok, I am doing it now.

File: libsolutil/StringUtils.h (hunk 16 above)
Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]: That is a good idea. We can never have enough tests.

File: libsolutil/StringUtils.h [hunk 18]
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
...
 
 	HexCase hexcase = HexCase::Mixed;
-	HexPrefix prefix = HexPrefix::Add;
+	HexPrefix hexprefix = HexPrefix::Add;
 
 	// when multiple trailing zero bytes, format as N * 2**x
 	int i = 0;
-	T v = _value;
-	for (; (v & 0xff) == 0; v >>= 8)
+	for (temp = signedValue; (temp & 0xff) == 0; temp >>= 8)
```
Comment (axic) [inline libsolutil/StringUtils.h]: I think it may be better scoping `temp` here and in the next loop, as they are not used inbetween.

Comment (axic) [reply libsolutil/StringUtils.h]: Would also keep it named `v` as before.

File: libsolutil/StringUtils.h [hunk 19]
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
...
 
 	// smaller numbers return as decimal
-	if (_value <= 0x1000000)
-		return _value.str();
+	if (signedValue <= 0x1000000)
+		return sign + signedValue.str();
 
 	HexCase hexcase = HexCase::Mixed;
-	HexPrefix prefix = HexPrefix::Add;
+	HexPrefix hexprefix = HexPrefix::Add;
```
Comment (axic) [inline libsolutil/StringUtils.h]: I'd also name these `hexCase` and `hexPrefix`.

File: libsolutil/StringUtils.h [hunk 20]
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
...
-		std::is_same<bigint, T>::value || !std::numeric_limits<T>::is_signed,
-		"only unsigned types or bigint supported"
-	); //bigint does not carry sign bit on shift
+		std::numeric_limits<T>::is_integer,
+		"only integer types are supported"
+	);
+
+	bool isNegative = _value < 0;
+	bigint temp;
+	bigint signedValue = isNegative ? (bigint(-1) * _value) : bigint(_value);
```
Comment (axic) [inline libsolutil/StringUtils.h]: Since this is copied, perhaps make it const. And `sign` too.

File: libsolutil/StringUtils.h (hunk 18 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: wouldn't using `temp` here make it more clear what the role of the variable is? It took me some time to understand what role `v` played in the code; and I think other programmers would face a similar situation when looking at the code.

Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]: I agree with naming the variable as `temp` and also with limiting its scope to only inside the loops.

Comment (Ruko97) [reply libsolutil/StringUtils.h]: ok, will do that

**Comment by Ruko97** [comment]
The errors in CircleCI seem to be related to not being able to find `fmt::format`. Does CircleCI not use the fmt library?

**Comment by cameel** [comment]
It's there but we added it as a dependency very recently so not all of our libs link to it yet. You have to add it to `target_link_libraries` in `libsolutil/CMakeLists.txt`. See how #13473 does it.

File: libsolutil/StringUtils.h [hunk 21]
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
...
+
+		// when multiple trailing zero bytes, format as N * 2**x
+		int i = 0;
+		for (temp = signedValue; (temp & 0xff) == 0; temp >>= 8)
+			++i;
+		if (i > 2)
+		{
+			// 0x100 yields 2**8 (N is 1 and redundant)
+			if (temp == 1)
+				return fmt::format("{}2**{}", sign, std::to_string(i * 8));
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
				return fmt::format("{}2**{}", sign, i * 8);
```
I guess we can remove the conversion to string now, since `fmt::format` should be able to handle it.

File: libsolutil/StringUtils.h [hunk 22]
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
...
+			else if ((temp & (temp-1)) == 0)
+			{
+				int j = 0;
+				for (; (temp & 0x1) == 0; temp >>= 1)
+					j++;
+				return fmt::format("{}2**{}", sign, std::to_string(i * 8 + j));
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
				return fmt::format("{}2**{}", sign, i * 8 + j);
```

File: libsolutil/StringUtils.h [hunk 23]
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
...
+			}
+			else
+				return fmt::format(
+					"{}{} * 2**{}",
+					sign,
+					toHex(toCompactBigEndian(temp), hexPrefix, hexCase),
+					std::to_string(i * 8)
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
					i * 8
```

File: libsolutil/StringUtils.h [hunk 24]
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
...
+		// when multiple trailing FF bytes, format as N * 2**x - 1
+		i = 0;
+		for (temp = signedValue; (temp & 0xff) == 0xff; temp >>= 8)
+			++i;
+		if (i > 2)
+		{
+			std::string suffix = isNegative ? " + 1" : " - 1";
+
+			if (temp == 0)
+				return fmt::format("{}2**{}{}", sign, std::to_string(i * 8), suffix);
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
				return fmt::format("{}2**{}{}", sign, i * 8, suffix);
```

File: libsolutil/StringUtils.h [hunk 25]
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
...
+			else if ((temp & (temp + 1)) == 0)
+			{
+				int j = 0;
+				for (; (temp & 0x1) != 0; temp >>= 1)
+					j++;
+				return fmt::format("{}2**{}{}", sign, std::to_string(i * 8 + j), suffix);
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
				return fmt::format("{}2**{}{}", sign, i * 8 + j, suffix);
```

File: libsolutil/StringUtils.h [hunk 26]
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
...
+			}
+			else
+				return fmt::format(
+					"{}{} * 2**{}{}",
+					sign,
+					toHex(toCompactBigEndian(++temp), hexPrefix, hexCase),
+					std::to_string(i * 8),
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
					i * 8,
```

File: libsolutil/StringUtils.h [hunk 27]
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
...
-		return str.substr(0, initialChars) +
-			"...{+" +
-			std::to_string(numSkipped) +
-			" more}..." +
-			str.substr(len-finalChars, len);
+		return fmt::format(
+			"{}{}...{{+{} more}}...{}",
+			sign,
+			str.substr(0, initialChars),
+			std::to_string(numSkipped),
```
Comment (matheusaaguiar) [inline libsolutil/StringUtils.h]:
```suggestion
			numSkipped,
```

**Review by matheusaaguiar — COMMENTED** [review]
Some adjustments and I wanted to confirm that you added the test case mentioned [here](https://github.com/ethereum/solidity/pull/13277#discussion_r978648348).
Also, @nikola-matic are you satisfied with the current replacements of string concatenation or do you think that they are needed for the simple cases at lines `133`, `203` and `218`?

**Comment by nikola-matic** [comment]
> Some adjustments and I wanted to confirm that you added the test case mentioned [here](https://github.com/ethereum/solidity/pull/13277#discussion_r978648348). Also, @nikola-matic are you satisfied with the current replacements of string concatenation or do you think that they are needed for the simple cases at lines `133`, `203` and `218`?

The formatting looks good now, and no, no need for `fmt::format` for simple cases. Another rebase would be nice to get rid of the `gp2` failure.

**Comment by Ruko97** [comment]
> Some adjustments and I wanted to confirm that you added the test case mentioned [here](https://github.com/ethereum/solidity/pull/13277#discussion_r978648348). Also, @nikola-matic are you satisfied with the current replacements of string concatenation or do you think that they are needed for the simple cases at lines `133`, `203` and `218`?

Yes, I added them at lines `185`, `186`, `285` and `286`. I added some more tests now just to make sure.

**Comment by Ruko97** [comment]
I am also removing the `std::to_string` functions, and rebasing the code to the latest version.

File: libsolutil/StringUtils.h (hunk 26 above)
Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]: This is still pending.

File: libsolutil/StringUtils.h (hunk 27 above)
Comment (matheusaaguiar) [reply libsolutil/StringUtils.h]: This is still pending.

File: libsolutil/StringUtils.h (hunk 26 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: fixed

File: libsolutil/StringUtils.h (hunk 27 above)
Comment (Ruko97) [reply libsolutil/StringUtils.h]: fixed

**Comment by nikola-matic** [comment]
@matheusaaguiar @ekpyron will this need a changelog entry?

**Comment by ekpyron** [comment]
> @matheusaaguiar @ekpyron will this need a changelog entry?

I'd say, it's borderline. But strictly speaking it's user-facing, so we can just default to "yes".

**Comment by matheusaaguiar** [comment]
@Ruko97 , please add a [Changelog](https://github.com/ethereum/solidity/blob/46a7ebd5b66c5638427295b9c3dbf15fc48c4ed2/Changelog.md) entry under the `Compiler Features` section, describing shortly the improvement on readability of large/hex numbers. Also, the fixed bug, under section `Bugfixes`.

File: Changelog.md [hunk 28]
```diff
@@ -8,9 +8,11 @@ Compiler Features:
 * Standard JSON: Add a boolean field `settings.metadata.appendCBOR` that skips CBOR metadata from getting appended at the end of the bytecode.
 * Yul Optimizer: Allow replacing the previously hard-coded cleanup sequence by specifying custom steps after a colon delimiter (``:``) in the sequence string.
 * Language Server: Add basic document hover support.
+* SMTChecker: Improved readability for large integers that are powers of two or almost powers of two
```
Comment (nikola-matic) [inline Changelog.md]: @leonardoalt is this SMT checker related?

**Comment by nikola-matic** [comment]
Rebased, resolved conflicts, and added periods to changelog entires. The build should now be fine, and hopefully closer to merging.

**Review by NunoFilipeSantos — APPROVED** [review]
LGTM

**Review by matheusaaguiar — APPROVED** [review]
LGTM
